ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+                    color,
+                    thickness // 3,
+                )
+
+        # Publish the image as a ROS message
+        self.viz_groundingdino_pub.publish(
+            cv2_image_to_ros_msg(image_copy, compress=False, bridge=self.bridge)
+        )
+
+    async def run_vision_pipeline(self, image_msg: Image, caption: str):
+        """
+        Run the vision pipeline consisting of two foundation models, GroundingDINO and
+        EfficientSAM, on the image. GroundingDINO is prompted with a caption and the latest image,
+        and outputs bounding boxes for each semantic label in the caption detected in the
+        image. The detected bounding boxes + semantic label pairs are then segmented by passing 
+        in the bounding box detections into EfficientSAM which outputs pixel-wise masks for each
+        bounding box. The top masks are then returned along with the semantic label for each mask 
+        as a dictionary.
+
+        Parameters
+        ----------
+        image: The image to segment, as a ROS image message.
+        caption: The caption to use for GroundingDINO containing all the food items
+                 detected in the image.
+
+        Returns
+        -------
+        result: The result message containing masks for all food items detected in the image
+                paired with semantic labels.
+        """
+        self._node.get_logger().info("Running the vision pipeline...")
+
+        # Set the initial time to measure the elapsed time running GroundingDINO on the
+        # desired image and text prompts.
+        inference_time = time.time()
+
+        # Define the result and create result message header
+        result = SegmentAllItems.Result()
+        result.header = image_msg.header
+        if self.camera_info is None:
+            self.camera_info = self._node.get_latest_msg(self.camera_info_topic)
+        if self.camera_info is not None:
+            result.camera_info = self.camera_info
+        else:
+            self._node.get_logger().warn(
+                "Camera info not received, not including in result message"
+            )
+
+        # Get the latest depth image and convert the depth image to OpenCV format
+        depth_img_msg = self._node.get_latest_msg(self.aligned_depth_topic)
+        depth_img = ros_msg_to_cv2_image(depth_img_msg, self.bridge)
+
+        # Convert the image to OpenCV format
+        image = ros_msg_to_cv2_image(image_msg, self.bridge)
+
+        # Run GroundingDINO on the image
+        bbox_predictions = self.run_grounding_dino(
+            image, caption, self.box_threshold, self.text_threshold
+        )
+
+        # Publish a visualization of the GroundingDINO predictions, if the visualization
+        # flag is set to true
+        if self.viz_groundingdino:
+            self.visualize_groundingdino_results(image, bbox_predictions)
+
+        # Collect the top contender mask for each food item label detected by
```

Comment:
I'd suggest having a parameter to the function (set by default to `1`) that controls how many of the top masks you use per food item